### ISSUE TITLE
With[] Function

### ISSUE THREAD

**Issue by rocky** [mathics/Mathics#1021]
[`With[]`](https://reference.wolfram.com/language/ref/With.html)  is needed to implement Combinatorica's `BinarySearch`. 

```
BinarySearch[l_?(Length[#] > 0&), k_?NumericQ, f_:Identity]:=
        With[{res = binarysearchchore[l, k, f]},
             res/; res =!= $Failed
        ]
```

Note: to get BinarySearch going, as a workaround I added the Combinatorica version 0.6 routine.

**Comment by mmatera** [comment]
I take this, but a first implementation would maybe consist of an alias for Block

**Comment by rocky** [comment]
@mmatera Thanks! Beggars can't be choosers so however you want to handle or whatever workarounds you want to switch to are fine by me. (I already put in my workaround, but if yours is better let's go with that.)

**Comment by mmatera** [comment]
#1025 seems to fix it in the proper way, as well as the Catch/Throw implementation

**Comment by rocky** [comment]
Ok. Along with that merge, let's also revert the Combinatorica code. 

BTW I just had another thought with respect to Combinatorica. The 0.6 package almost works with few changes. So maybe we will have two packages. 

And the CombinatoricaLite will pull in CombinatoricaV0.6.  So CombinatoricaLite will overwrite the CombinaticaV0.6 package with things that it does better.

**Comment by rocky** [comment]
Was implemented in 1.1.1
